sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -70,437 +75,361 @@ object AggregateUtils {
-        "DISTINCT and FILTER cannot be used in aggregate functions at the same time")
+/** Utility class to hold all of the metrics related to hash aggregation */
+case class GpuHashAggregateMetrics(
+    numOutputRows: GpuMetric,
+    numOutputBatches: GpuMetric,
+    totalTime: GpuMetric,
+    computeAggTime: GpuMetric,
+    concatTime: GpuMetric,
+    sortTime: GpuMetric,
+    spillCallback: RapidsBuffer.SpillCallback)
+
+/** Utility class to convey information on the aggregation modes being used */
+case class AggregateModeInfo(
+    uniqueModes: Seq[AggregateMode],
+    hasPartialMode: Boolean,
+    hasFinalMode: Boolean,
+    hasCompleteMode: Boolean)
+
+class GpuHashAggregateIterator(
+    cbIter: Iterator[ColumnarBatch],
+    groupingExpressions: Seq[Expression],
+    aggregateExpressions: Seq[GpuAggregateExpression],
+    aggregateAttributes: Seq[Attribute],
+    resultExpressions: Seq[NamedExpression],
+    childOutput: Seq[Attribute],
+    modeInfo: AggregateModeInfo,
+    metrics: GpuHashAggregateMetrics,
+    targetBatchSize: Long)
+    extends Iterator[ColumnarBatch] with Arm with AutoCloseable with Logging {
+  // Partial mode:
+  //  1. boundInputReferences: picks column from raw input
+  //  2. boundUpdateAgg: performs the partial half of the aggregates (GpuCount => CudfCount)
+  //  3. boundMergeAgg: (if needed) perform a merge of partial aggregates (CudfCount => CudfSum)
+  //  4. boundResultReferences: is a pass-through of the merged aggregate
+  //
+  // Final mode:
+  //  1. boundInputReferences: is a pass-through of the merged aggregate
+  //  2. boundMergeAgg: perform merges of incoming, and subsequent batches if required.
+  //  3. boundFinalProjections: on merged batches, finalize aggregates
+  //     (GpuAverage => CudfSum/CudfCount)
+  //  4. boundResultReferences: project the result expressions Spark expects in the output.
+  private case class BoundExpressionsModeAggregates(
+      boundInputReferences: Seq[GpuExpression],
+      boundFinalProjections: Option[scala.Seq[GpuExpression]],
+      boundResultReferences: scala.Seq[Expression],
+      aggModeCudfAggregates: scala.Seq[(AggregateMode, scala.Seq[CudfAggregate])])
+
+  Option(TaskContext.get()).foreach(_.addTaskCompletionListener[Unit](_ => close()))
+
+  private[this] val isReductionOnly = groupingExpressions.isEmpty
+  private[this] val boundExpressions = setupReferences(childOutput)
+  private[this] val aggregatedBatches = new util.ArrayDeque[LazySpillableColumnarBatch]
+  private[this] var outOfCoreIter: Option[GpuOutOfCoreSortIterator] = None
+
+  /** Iterator for fetching aggregated batches if a sort-based fallback has occurred */
+  private[this] var sortFallbackIter: Option[Iterator[ColumnarBatch]] = None
+
+  /** Whether a batch is pending for a reduction-only aggregation */
+  private[this] var hasReductionOnlyBatch: Boolean = isReductionOnly
+
+  override def hasNext: Boolean = {
+    // FIXME: totalTime accounting if it stays
```

Comment:
Yeah I'm not a fan of it either, which is why I didn't bother to plumb it through yet.  I left this as a marker to figure out what we really want to do.

@andygrove since you've been refactoring metrics lately for CBO, curious what metrics would be desired for this exec.

Reply:
The `cbIter` iterator could be wrapped in a `com.nvidia.spark.rapids.CollectTimeIterator` to automatically build `collectTime` metrics on each call to `next` and `hasNext`, if we want to measure that.

For CBO specifically, I would generally want to see metrics that we can directly compare to Spark's metrics so that we can produce a cost model comparing the two. For general profiling purposes, it is good to have metrics for the main operations being performed, like aggregate/merge/concat and it looks like we are already doing this.

Reply:
I think https://github.com/NVIDIA/spark-rapids/issues/2527 is the best guide for which metrics to add, except instead of `opTime` we would have multiple metrics for the different operations (aggregate/merge/concat) for these more complex operators.

Reply:
But do we need collect time or a total time for aggregations?  What metrics do we really need from aggregation?

Reply:
For CBO, I suppose we really just want the cost of this operator, exclusive of collectTime (unless collectTime is insignificant in which case we can just include it), so that we can estimate the cost of this operator for a given input row size, ignoring other factors that would affect execution time for now. If the individual metrics (aggregate, merge, etc) can be summed to give the total opTime, I think that would work.

For CPU we can use the optimizer operator isolation hacks to try and measure a total time for this operator (exclusive of child totalTime) and use that for the CPU estimate.